src/cassiopeia/critique/critique_utilities.py

```diff
@@ -179,3 +181,205 @@ def sample_triplet_at_depth(
         )
 
     return (str(in_group[0]), str(in_group[1]), str(out_group)), out_group
+
+
+def annotate_tree_depths_nx(G: nx.DiGraph) -> dict[int, list[Any]]:
+    """Annotates tree depth at every node for a networkx DiGraph.
+
+    Adds two attributes to the tree: how far away each node is from the root of
+    the tree and how many triplets are rooted at that node. Modifies the tree
+    in place.
+
+    Args:
+        G: nx.DiGraph
+
+    Returns:
+            A dictionary mapping depth to the list of nodes at that depth.
+    """
+    root = G.graph.get("root") or next((n for n in G.nodes if G.in_degree(n) == 0), None)
+    if root is None:
+        raise ValueError("No root found for depth annotation.")
+
+    # depth
+    depths = nx.single_source_shortest_path_length(G, root)
+    depth_to_nodes = defaultdict(list)
+    for node, depth in depths.items():
+        G.nodes[node]["depth"] = depth
+        depth_to_nodes[depth].append(node)
+
+    # leaf sizes
+    leaf_sz = {}
+    for u in reversed(list(nx.topological_sort(G))):
```

Comment:
Recalculating the number of leaves below each internal node is O(n^2). Can you come up with an O(n) way to do this? We can go over it tomorrow.